uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/app/IApp.java

```diff
@@ -83,4 +83,9 @@ public interface IApp {
      * Copy Installed APK to supplied destination.
      */
     void copyApk(String destApkFileName);
+
+    /**
+     * Determines the implementation to use based on the app version
+     */
+    void setAppImpl();
```

Comment:
I was actually hoping that it would be a "get" method rather than a "set"

Also I think this probably doesn't need to go in `IApp`, probably rather into the `App.java` as an `abstract` method. Or we can actually define a default implementation that just returns `this`

Reply:
then inside install you can assign as 

`mAppImpl = getAppImpl()`

Reply:
Shahzaib, can you elaborate on why it should be get method? My intention to write setter is to set the impl based on the version. I can move it to app.java instead of IApp.java but all methods like install, forceStop are currently in IApp.java. So i thought this would be consistent with other methods.

Reply:
The rest of the methods are inside `IApp` because those are actual functionalities by provided by definition of an app. Methods that are declared on an interface are `public` by nature and can be invoked by anyone from anywhere. `getAppImpl` method however, is sort of a custom functionality that I believe doesn't actually belong to an app definition, and therefore I think it shouldn't be publicly accessible i.e. not anyone should be able to call that method to get or set the App Impl. For this reason, I think we can move this method to `App.java` which is an abstract class and abstract classes can allow to declare both implemented as well as abstract methods with different visibility modifiers such as `public`, `protected` as well as `package-private`. I believe the `getAppImpl` method can be `protected` so that child classes can define their implementation of the impl, but only child classes can now see this method and no outsider class can see that method. I hope that makes sense.

Furthermore, regarding the "get" vs "set" - Setters generally accept an argument and set that. I don't see your method doing that so it is not really a setter. The way I see it, what your method is doing is really "getting or creating" an App Impl and then actually sets that impl somewhere as well, but I wouldn't call it a setter.

I mean a setter would be the following:

`public void setAppImpl(IApp appImpl) {
   mAppImpl = appImpl
}
`

Your method does the following however:
`public void setAppImpl() {
   mAppImpl = createAppImpl();
}`

So I think it can really refactor to the following:

```
class App {
  mAppImpl = getAppImpl(); // this can be renamed to createAppImpl

  protected App getAppImpl() {
     return this; // or create impl based on version for child classes
  }
}
```

Reply:
Shahzaib, I moved it to App.java and made it an abstract method in my local. But I wanted to close on the get/set part of it. If setAppImpl is giving the impression of a setter, then we can probably change it to initializeAppImpl. The issue I see with getAppImpl is that, I dont want to get it right away. I want to initialise the impl in install() and then when I call the individual methods like performDeviceRegistration, the initialised impl will simply be used. LMK if you think this could be an alternate solution?

Reply:
I also tried implementing this with generics like we discussed offline. But I hit a blocker, I must be knowing the type of implementation while creating an object of BrokerMicrosoftAuthenticator, which is not possible in this case. To keep this simple, I left it as is for now. I will give it another try in another PR seperately

Reply:
Sounds good